backend/src/routes/payments.js

```diff
@@ -0,0 +1,58 @@
+const express = require('express');
+const services = require('../services');
+const router = express.Router();
+const db = require('../db');
+const auth = require('../auth');
+const utils = require('../utils');
+
+const getPaymentRedirectUrl = async (req, res) => {
+    try {
+        const paymentObj = req.body;
+        const validationErrors = utils.payments.validateAmount(
+            paymentObj.amount
+        );
+        if (validationErrors) {
+            res.status(422).json(validationErrors);
+        } else {
+            const user = req.user;
+            const dbUser = await db.users.getUser(user.phoneNumber);
+            paymentObj.userId = dbUser.id;
+            paymentObj.username = dbUser.username;
+            const url = await services.payments.getPaymentRedirectUrl(
+                paymentObj
+            );
+            res.status(200).json(url);
+        }
+    } catch (err) {
+        res
+            .status(500)
+            .json(`Failed to get payment redirect url. Error: ${err.message}`);
+    }
+};
+
+const paymentNotify = async (req, res) => {
+    res.status(200).json('Payment notification request received from Bambora');
+};
+
+const paymentReturn = async (req, res) => {
+    const response = req.query.RETURN_CODE;
+    if (response === 0) {
+        const ordernumber = req.query.ORDER_NUMBER;
+        const payment = await db.payments.getPayment(ordernumber);
+        const dbUser = await db.users.getUserById(payment.userId);
+        const newBalance = payment.amount + dbUser.balance;
+        await db.reservations.updateUserBalance(dbUser.id, newBalance);
+        const balance = await db.reservations.getUserBalance(dbUser.id);
+        res
+            .status(200)
+            .json(`Payment successful. Updated user balance to: ${balance}`);
+    } else {
+        res.status(422).json(`Payment failed. Please try again later`);
+    }
+};
+
+router.post('/getPaymentRedirectUrl', auth.requireAuth, getPaymentRedirectUrl);
```

Comment:
Why is this a POST?

Reply:
?Oh, found it later. It is because we need to get it from Bambora side. Maybe we could rename this to reflect that operation like `generatePaymentRedirectUrl`?